### PR TITLE
Add feature to generate diffs for helm values.yaml file between releases

### PR DESCRIPTION
**Pull request by johnma14**
This feature adds the capability to compare changes to the configuration options
in the helm values.yaml files between the current and previous release. This was
included in the 1.1 release but did not get merged. Here is the link to the original PR: https://github.com/istio/istio.io/pull/3420

Fixes: #4384

### CODE REVIEW

File: scripts/tablegen.py [hunk 1]
```diff
@@ -332,11 +565,34 @@ def sanitizeValueStr(value):
...
+            # with the older version.
+            groups = re.search("\| \`(.*)\` \| \`(.*)\` \| (.*) |", value.strip())
+            if groups:
+                key = groups.group(1)
+                indexValue = indexMap.get(key)
+                
+                indexList.append(compareValues(indexValue, value, k))
         print ('')
-
+
```
Comment (howardjohn) [inline scripts/tablegen.py]: nit: extra space

**Comment by howardjohn** [comment]
Haven't had a change to go through the code yet. One thing that would be nice: If we the only thing we change is adding a description, we probably don't need to call that out. That seems to be most of the changes in https://istio.io/docs/reference/config/installation-options-changes/.

The changes are the most important thing there, because a user may implicitly depend on some value and we changed it, so we should make sure those don't get lost in the noise of us adding documentation.

Actually maybe just remove the descriptions entirely in the changed fields section. They are useful in the new fields section though so they don't have to go digging

File: scripts/tablegen.py (hunk 1 above)
Comment (johnma14) [reply scripts/tablegen.py]: will fix this. Thanks.

**Comment by johnma14** [comment]
@howardjohn That makes sense. I will go ahead and not track changes to the `Description` field in the `Modified` section but will retain it in the 'New' section. I will push a change later tonight with the latest changes. If you find any other changes that needs to be made to the installation_option_changes file, let me know. Thanks.

**Comment by rcaballeromx** [comment]
@johnma14 Can you help me understand how this PR relates to #4408? Does this script generate that page?

**Comment by howardjohn** [comment]
https://deploy-preview-4408--preliminary-istio.netlify.com/docs/reference/config/installation-options-changes/ shows one of the tables mangled. Search for `policy.autoscaleMin` and you should see, it is mid page

**Comment by johnma14** [comment]
@rcaballeromx yes this script generates the installation_options and installation_option_changes files. 

@howardjohn the files have been updated with the changes:
- If the only change is related to the description field, we do not show that as a change
- the mangled up options were removed after another clean run of the script.

**Comment by linsun** [comment]
@johnma14 thanks for updating this!  I agree with John that if the only change is desp, I'd rather not highlight it.

Could you ensure you are running the diff against 1.2?  Thanks!

**Comment by johnma14** [comment]
@linsun I made the changes so that if the change is only in the description field that it does not show it as a change. I will go ahead and run this script against the 1.2 release. I was actually running this against the master branch. Thanks Lin.

**Comment by sdake** [comment]
From latest render: https://deploy-preview-4402--preliminary-istio.netlify.com/docs/reference/config/installation-options/

Screen Shot 2019-06-15 at 3 14 53 AM

Can you set the correct version numbers in the md output?  This doesn't really need to be automated - something simple like manually modifying the MD template would be fine.

The output doesn't appear to be from the correct version.  see: https://deploy-preview-4402--preliminary-istio.netlify.com/docs/reference/config/installation-options-changes/]

For example, very few options have changed in the 1.2 release (vs the 1.1 release), so there should be very few new options.

**Comment by sdake** [comment]
can you please rebase on master to pick up the changes to fix circleci?

Thanks!
-steve

**Comment by sdake** [comment]
> @linsun I made the changes so that if the change is only in the description field that it does not show it as a change. I will go ahead and run this script against the 1.2 release. I was actually running this against the master branch. Thanks Lin.

perhaps one part of the diff was master, but the source looks like it was against 1.0, instead of 1.1?

**Comment by sdake** [comment]
@johnma14 hi - We are shipping 1.2 in a few days (Tuesday iirc).  Are you able to get this PR merged prior to then against the latest rc?  Once 1.2 is released, a new PR will be required against the 1.2 tarballs to ensure all the tags and hubs are correct.

Cheers
-steve

**Comment by johnma14** [comment]
@sdake yes I think we can get this done before that. I am going to push out the latest change and we can review and verify. Thanks @sdake.

**Comment by johnma14** [comment]
@sdake @linsun @howardjohn I just pushed a new change (here: https://github.com/istio/istio.io/pull/4408) for the installation-options and installation-option-changes to reflect installation options in the release-1.2.0-rc-2. I cross-checked the values between the release and what's in the new report and it looks right to me. Please let me know if I have missed anything or it needs any update and I will fix it. Thanks, appreciate it.

**Comment by geeknoid** [comment]
So what should we do with this PR? Should we merge to master and cherry pick to the 1.2 branch?

**Comment by johnma14** [comment]
@geeknoid We have been using this to generate the diffs for 1.0->1.1 and 1.1->1.2. This did not get merged into 1.1 but I think we should definitely merge this into the master. Whether we cherry-pick to 1.2 I do not know. If we know for sure that there will be no configuration option changes in 1.2, then we probably do not need to cherry-pick this to 1.2, but I am not sure. Thanks.

**Comment by sdake** [comment]
@geeknoid I have been on semi-vacation for 6 weeks, so been putting off the hard reviews :)

I finallly had a chance to speak with @johnma14 on Slack today, as I was struggling to understand the code flow.  The root of my problem was that tablegen and diff generation were rolled into one python tool.  Instead of that approach, @johnma14 is going to submit this work under a different binary name, and then we will backport that work.

The main reason for this is I want to ensure `tablegen.py` could potentially be used for the next-gen installer in 1.3 or 1.4 (or when it lands) assuming the protobufs documentation lands in the long term (IOW this may be a gap solution or it may be a long term solutino).

here is the slack log from Istio's @environments slack channel:
```
Steve Dake [11:50 AM]
@mariam john hi , just working through my backlog of reviews after vvacation, not sure if your around, if you are, ping me in #environments


mariam john [11:51 AM]
was added to #environments by Steve Dake.

mariam john [11:51 AM]
Thanks @sdake

Steve Dake [11:51 AM]
hi
re review:

mariam john [11:52 AM]
https://github.com/istio/istio.io/issues/4384
johnma14
#4384 Generate diffs for helm values.yaml file for 1.1 -> 1.2
Generate the installation option changes for release 1.1 to 1.2.
Assignees
johnma14
Labels
kind/docs
istio/istio.io | Jun 13th | Added by GitHub

Steve Dake [11:53 AM]
https://github.com/istio/istio.io/pull/4402
johnma14
#4402 Add feature to generate diffs for helm values.yaml file between releases
This feature adds the capability to compare changes to the configuration options  
in the helm values.yaml files between the current and previous release. This was  
included in the 1.1 release but did not get merged. Here is the link to the original PR: #3420

Fixes: #4384
Assignees
johnma14
Labels
cla: yes
istio/istio.io | Jun 13th | Added by GitHub
I was looking at this one

mariam john [11:53 AM]
We have been using this script to generate the installation options and the installation option diffs for 1.1 and 1.2 releases
yes

Steve Dake [11:53 AM]
ok - which one should I look at, are they duplicates? :slightly_smiling_face:

mariam john [11:53 AM]
yeah they are duplicates actually

Steve Dake [11:54 AM]
can you close the one you dont want peeps to look at
and i’ll review the open one

mariam john [11:54 AM]
we should close one. let me check which one has the PR attached
sure
alright. closed the 4384 one

Steve Dake [11:55 AM]
we typically shouldn’t do reviews on slack - git hub is better for that, but for htis one in particular I need a little bit more context and I knwo you have been waiting around for months

mariam john [11:56 AM]
sure, 4402 is a good one to review. That has the changes we want to review

Steve Dake [11:56 AM]
reading the code
it sort of looks like values.yaml is processed for comments, and then the diff is produced from the outputs of those vaues.yaml: https://github.com/istio/istio.io/pull/4402/files#diff-01cb53a05c28022ae45a91fa1727c0eeR587
is this accurate?

mariam john [11:57 AM]
that is right.

Steve Dake [11:57 AM]
(the code is hard to follow, its my fault, I wrote it about a year ago and it was pretty hard to write back then :)
so you need 2 inputs, the old values.yaml and the new values.yaml?

mariam john [11:57 AM]
we parse the values.yaml to generate the installation options which also include the documentation for these options. and from that we generate the diffs
we generate the installation options from the values.yaml from the previous release and the new release and output it in the md format and then compare the 2
so yes it needs 2 inputs to generate the diffs
1) In this code section: in this code section: https://github.com/johnma14/istio.io/commit/f8bd75b9bd6d8470695b6807eac5f92a746838ed#diff-01cb53a05c28022ae45a91fa1727c0eeR557,
we parse the content from the previous release and store in a dictionary called indexMap. example: {certmanager.enabled: false,
         certmanager.replicaCount: 1}
2) In the code section: https://github.com/johnma14/istio.io/commit/f8bd75b9bd6d8470695b6807eac5f92a746838ed#diff-01cb53a05c28022ae45a91fa1727c0eeR579,
we generate the index.md for the current release and iterate through these options, we check if these options are available in the previous release and compare them to see what the changes are.

Steve Dake [12:02 PM]
i’m not super keen on generating the diffs alongside the table generation, but it is what it is, and I think can be cleaned up later if need be
since its simply a tool for generating docs
(rather than part of the main execution path)_
what is the workflow?

mariam john [12:03 PM]
We can decouple those if that is how we want to do it. That should not be too hard to do
https://github.com/johnma14/istio.io/commit/f8bd75b9bd6d8470695b6807eac5f92a746838ed#diff-01cb53a05c28022ae45a91fa1727c0eeR345 is the function that compares the old value and new value and tracks the differences in ordered dictionaries defined here: https://github.com/johnma14/istio.io/commit/f8bd75b9bd6d8470695b6807eac5f92a746838ed#diff-01cb53a05c28022ae45a91fa1727c0eeR79

Steve Dake [12:05 PM]
be back in 10 mins, need to collect daughter from summer school

mariam john [12:05 PM]
so we track the options that were just modified, the new ones that were added in the new release and the old ones that are removed in the new release
sure,  no worries. I will just leave some comments in here (whenever you have some time to review it)
3) This section: https://github.com/johnma14/istio.io/commit/f8bd75b9bd6d8470695b6807eac5f92a746838ed#diff-01cb53a05c28022ae45a91fa1727c0eeR607,
writes all the changes tracked in the ordered dictionaries (new ones added, old one removed and the modified changes - we only track the changes to the configuration option values and not the changes to the description) to the index.md file that tracks the diffs.

Steve Dake [12:17 PM]
back
ok - so the diffs I am less concerned about
thats a cool feature, but want to understand what files I put where to make the thing run? :slightly_smiling_face:
@mariam john about?

mariam john [12:21 PM]
right. So this depends on istio directory being under the istio.io directory in the dev env

Steve Dake [12:21 PM]
so ~/go/src/istio.io/istio

mariam john [12:21 PM]
right

Steve Dake [12:21 PM]
thats where its supposed to go :slightly_smiling_face:
kind of assuming only maintainers will run this code, and have their dirs in the correct location

mariam john [12:22 PM]
it gets the previous release from https://api.github.com/repos/istio/istio.io/branches
although that url doesn’t show the 1.1 or 1.2 release now

Steve Dake [12:23 PM]
does it base the previous release on the 1.1 release?

mariam john [12:23 PM]
not sure how to get those releases to show up there.
yes so that github api lists all the branches for istio and we get the latest from that list.
so before 1.2 was released, the highest branch there was the 1.1

Steve Dake [12:25 PM]
i recommend using https://api.github.com/repos/istio/istio/branches
ok - so it gets the prior release, then what ?

mariam john [12:25 PM]
alright. Will do that
we parse the content from the previous release and store in a dictionary called indexMap. example: {certmanager.enabled: false,
        certmanager.replicaCount: 1}

Steve Dake [12:27 PM]
parse the content from the values.yaml files?

mariam john [12:27 PM]
from the index.md of the previous release (which was generated from the values.yaml file)

Steve Dake [12:28 PM]
when does the values.yaml file create an index.md?

mariam john [12:30 PM]
there is this part in the code: https://github.com/istio/istio.io/blob/master/scripts/tablegen.py#L97
scripts/tablegen.py:97
```def process_helm_yaml(values_yaml, option):```
istio/istio.io | Added by GitHub

Steve Dake [12:30 PM]
so - we have 2 versions, 1.1, 1.2 - the way I think the code would work is values.yaml is pulled for release-1.1, values.yaml is pulled for release-1.2, the values.yaml files are serialized into yaml
internal python yaml  that is, and then a diff is produced
if that isn’t how it works, thats fine
I just want to understand what you ahve :slightly_smiling_face:
the first and foremost thing that must be correct is the installation-options page
that must always be correct, so want to make sure it doesn’t end corrupted by any of this diffing going on

mariam john [12:32 PM]
The reason why I am not looking at the values.yaml for the previous release is because it is not just 1 values.yaml we parse to generate the installation options for that release. So it was really an attempt to make things faster - relying on the assumption that the index.md for the previous release was stable and would not change.

Steve Dake [12:32 PM]
correct > speed
ok - so back on task - the values.yaml is looked at in the prior relese, what bootstraps the whole thing?

mariam john [12:33 PM]
Like you said , if we decouple the generation of installation options and installation option diffs. then we should be able to this kind of processing where we go through the values.yaml for the current and previous release and compare them. We can do that

Steve Dake [12:34 PM]
I am thinking about the next-gen installer mostly

mariam john [12:34 PM]
yeah that makes sense

Steve Dake [12:34 PM]
although I’d like the pr for what you have to land - I am curious how the index.md for the main options page is generated

mariam john [12:35 PM]
you mean the installation options?

Steve Dake [12:35 PM]
yhes

mariam john [12:36 PM]
that is mainly done by the code that you wrote - this section in particular - https://github.com/istio/istio.io/blob/master/scripts/tablegen.py#L97

Steve Dake [12:36 PM]
what inputs rae used for that though?

mariam john [12:36 PM]
the only change I made to that part is -
in the old version, it was only looking at the values.yaml in the charts directory, whereas in the newer version we look at the values.yaml in the charts directory and all the subcharts directory
plus we also added additional code to parse the description for each configuration option and populate them as well

Steve Dake [12:38 PM]
ok - so master version looks fine

mariam john [12:38 PM]
In this section: https://github.com/istio/istio.io/blob/master/scripts/tablegen.py#L72 , you can see how we iterate through the subcharts directory
scripts/tablegen.py:72
```def decode_helm_yaml(s):```
istio/istio.io | Added by GitHub

Steve Dake [12:39 PM]
this does the MD output after yaml serialization: https://github.com/istio/istio.io/blob/master/scripts/tablegen.py#L312-L342
scripts/tablegen.py:312-342
```with open(os.path.join(ISTIO_IO_DIR, CONFIG_INDEX_DIR), 'r') as f:
    endReached = False

    data = f.read().split('\n')
    for d in data:```
 Show more
istio/istio.io | Added by GitHub
got it

mariam john [12:39 PM]
thats right

Steve Dake [12:39 PM]
so the new version works the same way?

mariam john [12:39 PM]
yes it does

Steve Dake [12:39 PM]
ok - and it pulls the md from istio.io release-1.1?

mariam john [12:40 PM]
that’s right

Steve Dake [12:40 PM]
got it
ok - what I’d suggest is to decouple the table generation from the table diffing

mariam john [12:40 PM]
ok, should be able to do that

Steve Dake [12:40 PM]
but I dont wan tyou to spend a bunch of time on rework
is it a bunch of time?
nice improvementso n tablegen.py btw, the first version was bad, the master version is good !

mariam john [12:41 PM]
should not take a whole lot but I do have some things on my plate that I am already working on - I can try to get this done by the end of the week

Steve Dake [12:42 PM]
ok - well lets do this - please rename your binary to something other then tablegen.py
we can merge it as is, assuming it works

mariam john [12:42 PM]
ok

Steve Dake [12:42 PM]
call is diffgen.py or choose your own adventure :slightly_smiling_face:
this forces a decoupling

mariam john [12:43 PM]
will do that. I will test it and push another change

Steve Dake [12:43 PM]
just push in the current change please
git mv if you like
otherwise context is lost

mariam john [12:43 PM]
ok will do

Steve Dake [12:43 PM]
rockin

mariam john [12:43 PM]
thanks @sdake. appreciate you taking the time

Steve Dake [12:44 PM]
thanks @mariam john - then we can merbge this and backport
re backport

mariam john [12:44 PM]
sure

Steve Dake [12:44 PM]
this should only go to the 1.2 branch correct?
or not backported at all?

mariam john [12:44 PM]
we used this same code to generate 1.1 diff with 1.0 as well.

Steve Dake [12:45 PM]
cool
so backport it is :slightly_smiling_face:

mariam john [12:45 PM]
ok
```

**Comment by sdake** [comment]
> #4408

@rcaballeromx ya, this had me confused too.  This PR produces a list of differences, the other PR you referenced is a table of the installation options.  The code was in the same python binary, so this caused serious confusion for me.

The diff producer (this code) should be a separate concern (and for now a separate binary).  Our primary objective is to ensure the installation options table is of high quality (which needs some work on the values.yaml docificaiton).  This diff is helpful for operators evaluating Istio or upgrading from one version to another.

We need better installation option hygiene - including documentation as you rightly point out in other reviews.

Cheers
-steve

**Comment by rcaballeromx** [comment]
@sdake Thanks for clarifying.

**Comment by johnma14** [comment]
I made the following changes to the latest PR (mainly separated the code that generates installation options for the current release and the installation option differences between the current release and previous release): 
- Added some comments to tablegen.py to note that this file is used to generate Installation Options and fixed a minor typo in the existing comments section
- Added a new file tablegen_diff.py which generates the difference between the installation options for the current release and previous release. 

Thank you.

**Comment by johnma14** [comment]
@sdake @geeknoid @linsun @howardjohn @rcaballeromx
I made the following changes to the latest PR (mainly separated the code that generates installation options for the current release and the installation option differences between the current release and previous release): 
- Added some comments to tablegen.py to note that this file is used to generate Installation Options and fixed a minor typo in the existing comments section
- Added a new file tablegen_diff.py which generates the difference between the installation options for the current release and previous release. 

Thank you.

**Comment by sdake** [comment]
Screen Shot 2019-07-14 at 7 00 35 AM

The gate is blocked by a maintenance window.  This PR is probably safe to merge anyway.

Cheers
-steve

**Comment by sdake** [comment]
@geeknoid can you advise how to merge this PR?  Looks like mergify is in some wonky state.
